File: leetcode/explore/binary tree/postorder_traversal.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        nodeStack = [root]
        result = []
        while nodeStack:
            cur = nodeStack.pop()
            result.append(cur.val)
            if cur.left:
                nodeStack.append(cur.left)
            if cur.right:
                nodeStack.append(cur.right)
        return result[::-1]
    
    def postorderTraversal_recursive(self, root: TreeNode) -> List[int]:
        nodeStack = []
        self.helper(root, nodeStack)
        return nodeStack
    

    def helper(self, root: TreeNode, nodeStack: List[int]):
        if not root:
            return
        
        self.helper(root.left, nodeStack)
        self.helper(root.right, nodeStack)
        nodeStack.append(root.val)


if __name__ == "__main__":
    tn = TreeNode(1)
    cur = tn
    cur.right = TreeNode(2)
    cur = cur.right
    cur.left = TreeNode(3)
    print(Solution().postorderTraversal(tn))
    print(Solution().postorderTraversal_recursive(tn))

File: leetcode/explore/binary tree/postorder traversal.cpp
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        if (!root) return {};

        stack<TreeNode*> nodeStack;
        vector<int> result;
        nodeStack.push(root);

        while (!nodeStack.empty())
        {
            auto curNode = nodeStack.top();
            nodeStack.pop();
            result.push_back(curNode->val);
            if (curNode->left) nodeStack.push(curNode->left);
            if (curNode->right) nodeStack.push(curNode->right);
        }
        reverse(result.begin(), result.end());
        
        return result;
    }

    vector<int> postorderTraversal_recursive(TreeNode* root) {
        vector<int> result;
        helper(root, result);
        return result;
    }

    void helper(TreeNode* root, vector<int>& result) {
        if (!root) return;
        helper(root->left, result);
        helper(root->right, result);
        result.push_back(root->val);
    }
};


int main() {
    TreeNode* root = new TreeNode(1);
    auto cur = root;
    cur->right = new TreeNode(2);
    cur = cur->right;
    cur->left = new TreeNode(3);
    Solution s;
    s.postorderTraversal(root);
}